src/gwt/src/org/rstudio/studio/client/application/ui/CommandPalette.java

```diff
@@ -0,0 +1,442 @@
+            continue;
+         }
+         
+         // Ensure the command can be used. It'd be nice to show all commands in
+         // the palette for the purposes of examining key bindings, discovery,
+         // etc., but there's no good user experience if a user attempts to
+         // invoke one of those commands.
+         AppCommand appCommand = allCommands.get(command);
+         if (!appCommand.isEnabled() || !appCommand.isVisible())
+         {
+            continue;
+         }
+
+         // Look up the key binding for this command
+         List<KeySequence> keys = map.getBindings(command);
+         
+         // Create an application command entry
+         CommandPaletteEntry entry = new AppCommandPaletteEntry(appCommand, keys);
+         if (StringUtil.isNullOrEmpty(entry.getLabel()))
+         {
+            // Ignore app commands which have no label
+            continue;
+         }
+         entries_.add(entry);
+      }
+      
+      // Add all of the R addin commands
+      map = shortcuts_.getKeyMap(KeyMapType.ADDIN);
+      AddinExecutor executor = new AddinExecutor();
+      for (String addin: JsUtil.asIterable(addins_.keys()))
+      {
+         RAddin rAddin = addins_.get(addin);
+         
+         // Look up the key binding for this addin
+         List<KeySequence> keys = map.getBindings(rAddin.getId());
+         CommandPaletteEntry entry = new RAddinCommandPaletteEntry(rAddin, executor, keys);
+         if (StringUtil.isNullOrEmpty(entry.getLabel()))
+         {
+            // Ignore addin commands which have no label
+            continue;
+         }
+         entries_.add(entry);
+      }
+      
+      // Invoke commands when they're clicked on
+      for (CommandPaletteEntry entry: entries_)
+      {
+         entry.sinkEvents(Event.ONCLICK);
+         entry.addHandler((evt) -> {
+            host_.dismiss();
+            entry.invoke();
+         }, ClickEvent.getType());
+         commandList_.add(entry);
+      }
+
+      
+      // Handle most keystrokes on KeyUp so that the contents of the text box
+      // have already been changed
+      searchBox_.addKeyUpHandler((evt) ->
+      {
+         if (evt.getNativeKeyCode() == KeyCode.ESC)
+         {
+            // Pressing ESC dismisses the host (removing the palette popup)
+            host_.dismiss();
+         }
+         else if (evt.getNativeKeyCode() == KeyCode.ENTER)
+         {
+            // Enter runs the selected command
+            invokeSelection();
+         }
+         else
+         {
+            // Just update the filter if the text has changed
+            String searchText = searchBox_.getText();
+            if (!StringUtil.equals(searchText_, searchText))
+            {
+               searchText_ = searchText;
+               applyFilter();
+            }
+         }
+      });
+
+      // Up and Down arrows need to be handled on KeyDown to account for
```

Comment:
As mentioned elsewhere, support for PgUp and PgDown keys would be really nice here, too.

Reply:
Wish granted! 🧞